Script functions: add support for inline translatable documentation

zas opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Picard. We
    appreciate your time and interest in helping our project!

    Use this template to help us review your change. Not everything is required,
    depending on your change. Keep or delete what is relevant for your change.
    Remember that it helps us review if you give more helpful info for us to
    understand your change.

    Ensure that you've read through and followed the Contributing Guidelines, in
    [CONTRIBUTING.md](https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md).
-->

# Summary

<!--
    Update the checkbox with an [x] for the type of contribution you are making.
-->

* This is a…
    * [ ] Bug fix
    * [x] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem

Current script functions documentation is maintained outside the code, and as html document.

<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

* JIRA ticket (_optional_): PICARD-XXX
<!--
    Please make sure you prefix your pull request title with 'PICARD-XXX' in order
    for our ticket tracker to link your pull request to the relevant ticket.
-->



# Solution

Use markdown and gettext, along strings stored next to function declarations to ease script function documentation and translation.

A part of this PR is dedicated on changing interface (script documentation is now displayed in-app).
Also since functions are listed, script editor completion is now making use of them.

<!--
    The details of your change. Talk about technical details, considerations, or
    other interesting points. If you have a lot to say, be more detailed in this
    section.
-->


# Action

<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->

Sophist-UK commented:
Nice.

zas commented:
Since Qt 5.14  there are new methods to display markdown: https://doc.qt.io/qt-5/qtextdocument.html#setMarkdown

zas commented:
Still very experimental, but here we go:

![Capture du 2020-03-09 10-49-34](https://user-images.githubusercontent.com/151042/76201803-217d5d80-61f4-11ea-8591-257cbf640f08.png)

zas commented:
![image](https://user-images.githubusercontent.com/151042/76230292-c95e4f80-6223-11ea-9605-3de5c9e9e2b8.png)

zas commented:
![image](https://user-images.githubusercontent.com/151042/76279540-33600e80-628f-11ea-9bb8-9f4f8bf7c51b.png)

zas commented:
For some reason geometry of the dialog isn't restored after closing and re-opening it.

zas commented on picard/ui/options/scripting.py:
```diff
@@ -166,12 +166,17 @@ def __init__(self, parent):
             except ValueError:
-                return '<dt>$' + function.name + '(...)</dt><dd>' + html + '</dd>'
+                return template % ("$%s()" % function.name, module, html)
 
```
A plugin script function can be documented as:

```python
@script_function(documentation="test")
def keep():
   code...
```
![image](https://user-images.githubusercontent.com/151042/76402599-ef9b0100-6383-11ea-85f0-10660e8012a9.png)

zas replied:
Note, in the image above, there's also a translation test (it works!).

zas commented:
I moved the external link to the dialog, it gives more room in script editor dialog.

![image](https://user-images.githubusercontent.com/151042/76456577-7a591b80-63d7-11ea-9613-5421e16497c4.png)

rdswift commented:
I like what you've done in adding the documentation into Picard directly so that it can be maintained in one single place and displayed within the program, but that raises a couple of questions for me:

1. Where does this place the work that I've been doing to restructure the function documentation into separate detailed web pages (including examples) for each function?  It seems that might now be a waste of time, in which case I'll cease work on it and delete the repository.

2. Should the examples that I've put together for each function be added into the descriptions in the code base?  If not, is there a plan as to where they should appear (if at all), and how to get them there?

3.  Is there an intent to provide documentation outside of the program which allows quick lookup of a specific function or finding (browsing) functions within function groups (e.g.: assignment, information, conditional, multi-value, etc.)?  This would be useful for those that develop their scripts outside of Picard using other editors.  I assume that the Picard website will still provide all of the functions listed (without examples or grouping?) on a single web page.

I know this sounds like whining, but I'm only trying to get an idea of where I should (and more importantly should not) be directing my efforts.  Thanks.

zas commented:
> * Where does this place the work that I've been doing to restructure the function documentation into separate detailed web pages (including examples) for each function?  It seems that might now be a waste of time, in which case I'll cease work on it and delete the repository.
> 

We'll be able to generate script functions documentation as one or more pages, that's not done yet, but the internal display is already html.

> * Should the examples that I've put together for each function be added into the descriptions in the code base?  If not, is there a plan as to where they should appear (if at all), and how to get them there?
> 

Perhaps we could store examples in a different field than `documentation`, like a list of markdown texts.
So we could more easily split documentation and examples.

> * Is there an intent to provide documentation outside of the program which allows quick lookup of a specific function or finding (browsing) functions within function groups (e.g.: assignment, information, conditional, multi-value, etc.)?  This would be useful for those that develop their scripts outside of Picard using other editors. 
> I assume that the Picard website will still provide all of the functions listed (without examples or grouping?) on a single web page

The idea is to make Picard website auto update Picard git repository changes, so it could keep the online documentation up-to-date.
We plan to develop a tool to manage that, based on changes made in this PR.
For now, we focus on getting it right for the in-app display and for translations.

zas commented:
>     2\. Should the examples that I've put together for each function be added into the descriptions in the code base?  If not, is there a plan as to where they should appear (if at all), and how to get them there?

I was thinking about something like:

![image](https://user-images.githubusercontent.com/151042/76669112-85a07880-658a-11ea-9f22-b9ca2bfba8a9.png)

We can pass extra arguments to registering function (or decorator), I did an attempt with:

```python
@script_function(eval_args=False, documentation=N_(
    """`$noop(...)`

Does nothing (useful for comments or disabling a block of code)."""
), examples=(
    """$noop()""",
    """$noop(this is ignored)"""
))
def func_noop(parser, *args):
    return ''
```

The idea is to split function description and its examples, examples is a list of markdown texts.

I think we should make examples actual working code, so we can test them (being parseable at least).

rdswift commented:
> The idea is to split function description and its examples, examples is a list of markdown texts.

I agree they should be separate from the description, and markdown is as good as any.

> I think we should make examples actual working code, so we can test them (being parseable at least).

That's probably a good idea, but if we're illustrating (and clarifying / explaining) the use of the functions, we should show the statement and the result.  See the examples for the [$substr function](https://rdswift.github.io/picard-docs/functions/func_substr.html) to see what I mean.  I suppose this could be handled by entering the examples as a list of tuples, each containing the code and the result, or some other way if you prefer.

What I was sort of mapping out in my head (instead of sleeping for the last few nights) is a system with each component entered separately and combined as required.  I was toying with the idea of a `FunctionInformation` object containing:
* function_name
* short_description (for a simple hyperlink list / table of functions)
* long_description
* picard_version_implemented
* examples
* arguments (list of `FunctionArgument` objects)

The `FunctionArgument` object could contain:
* argument_name
* type (e.g.: "string", "integer", "multi-value")
* required (is the argument required)
* multiple (can there be multiple arguments)
* default_value
* nullable (can the value be "" or "0")

This would allow the "Usage:" string to be built automatically, and the various help and function list pages to be built (and consistently formatted) automatically.  It could also allow for a class method to provide some basic testing of the validity of the arguments when the function is called.

I hadn't started coding any of it yet (probably a good thing) so there has been no thought as to how it could be made "developer friendly" or how it could be optimized (e.g.: combining "required" and "default_value").  Just a concept at this point.

rdswift commented:
After thinking about it this evening, I doubt that the flexibility provided by the system that I conceptualized in my last comment would ever be used, thus there is no value in creating such a system.  I'm sorry I even brought it up.

However, based on the great number of basic scripting questions in the Community Forum, I do believe that there is value in having documentation that is more geared to the novice or non-programmers that want to do a bit more with Picard to match their specific needs and wants.  Sort of a "Picard for Dummies" or "Introduction to Developing Scripts for Picard" manual.  Something with more extensive descriptions of the functions, and a number of well documented practical usage examples, with perhaps even some sort of tutorial (or tutorials) to help the novices that want to learn how to do things for themselves.  I think that goes far beyond what documentation is (and should be) captured in the code.  I believe that it really is a separate project.  The trick will be in developing something such that it does not compete with the current Picard documentation, but can build on it and augment it without duplicating all the work that's going into this PR.

I'm a firm believer in entering information once and then set up one or more systems to extract, combine, rearrange, format and otherwise manipulate the information to build the desired end products.  The questions then become:

1. What information is required?
2. What information is stored where?
3. How are the different information sources linked?
4. How are the sources accessed to provide the necessary combined information to meet a need?

Unfortunately, I suspect that the end products differ so greatly that implementing and maintaining such an information system will take more effort than it would take to address the end products separately.  Consequently, if you don't feel that examples are required for all of the functions (since they are not in the current documentation), then they probably shouldn't be added to the documentation in the code (at this time).  There's no value in putting them in if there is no plan to use them.

zas commented:
The idea for now is to round corners of what was made until now (in-app script function documentation based of translatable strings, next to related code, and completion of function names in script editor).

Another option for examples: display them on website, and link to them in the in-app documentation (since external links are allowed, and open in local browser when needed).

zas commented on picard/script.py:
```diff
@@ -145,9 +151,81 @@ def eval(self, state):
+        if self.documentation is not None and markdown is not None:
+            ret = markdown(_(self.documentation))
+        else:
+            ret = ''
```
Not sure what you mean, I think it would change the expected behavior.

phw replied:
My thought was to remove duplication. We already have `markdowndoc` which deals with returning 

```
if self.documentation is not None:
            ret = _(self.documentation)
```

This code is duplicated in `htmldoc`. My example however changed how the `preprocessor` is applied. So what about this instead:

```
def htmldoc(self, preprocessor=None):
    if markdown is not None:
        ret = markdown(self.markdowndoc())
    else:
        ret = ''
    return self._preprocess(ret, preprocessor)
```

This should be functionally equivalent. On a side note I think `preprocessor` should be renamed to `postprocessor`.

zas replied:
Ah yes, it makes sense, I did suggested changes in c3584da and 6879625

zas commented:
> Also, can we merge this into the devel branch for now?

I rebased the PR on devel (and moved to devel as basis).

>  I think we should make a 2.3.2 release first for bug fixes and then start merging everything for 2.4

I agree.

zas commented:
>  What I don't quite like is the position of the "Scripting Documentation" button, that looks odd.

Not happy with it either. Feel free to fix to your taste.
